torch/_inductor/fx_passes/b2b_gemm.py

```diff
@@ -0,0 +1,206 @@
+)
+
+
+def can_apply_b2b_gemm(
+    mat1: torch.fx.Node, mat2: torch.fx.Node, mat3: torch.fx.Node
+) -> bool:
+    if not (("val" in mat1.meta) and ("val" in mat2.meta) and ("val" in mat3.meta)):
+        return False
+    mat1 = mat1.meta["val"]
+    mat2 = mat2.meta["val"]
+    mat3 = mat3.meta["val"]
+    if not (mat1.is_cuda and mat2.is_cuda and mat3.is_cuda):
+        return False
+    if not (
+        (len(mat1.shape) == 2) and (len(mat2.shape) == 2) and (len(mat3.shape) == 2)
+    ):
+        return False
+    if not ((mat1.shape[1] == mat2.shape[0]) and (mat2.shape[1] == mat3.shape[0])):
+        return False
+    # TODO: change to a real-check for size restrictions (may consider hardware limit?)
+    m, n, o, p = mat1.shape[0], mat1.shape[1], mat3.shape[0], mat3.shape[1]
+    m_ok = (m % 128 == 0) and (m > 128)
+    n_ok = n == 32
+    o_ok = (o % 128 == 0) and (o > 128)
+    p_ok = p == 32
+    return m_ok and n_ok and o_ok and p_ok
+
+
+def tuned_b2b_gemm(mat1, mat2, mat3, *, layout=None):
+    layout = FixedLayout(
+        mat1.get_device(), mat1.get_dtype(), [mat1.shape[0], mat3.shape[1]]
+    )
+    choices: list[TritonTemplateCaller] = []
+    # TODO: change N and P to non-constexpr
+    # Note: the only reason why N and P are hardcoded is because in Triton tl.arange(0, N) only works for tl.constexpr
+    # TODO: add more configs for tuning (shall I also tune num_stages and num_warps?)
+    for config in [
+        {
+            "ROW_BLOCK_SIZE": 32,
+            "COL_BLOCK_SIZE": 32,
+            "num_stages": 2,
+            "num_warps": 4,
+            "N": 32,
+            "P": 32,
+        },
+        {
+            "ROW_BLOCK_SIZE": 32,
+            "COL_BLOCK_SIZE": 128,
+            "num_stages": 2,
+            "num_warps": 4,
+            "N": 32,
+            "P": 32,
+        },
+        {
+            "ROW_BLOCK_SIZE": 128,
+            "COL_BLOCK_SIZE": 32,
+            "num_stages": 2,
+            "num_warps": 4,
+            "N": 32,
+            "P": 32,
+        },
+        {
+            "ROW_BLOCK_SIZE": 128,
+            "COL_BLOCK_SIZE": 128,
+            "num_stages": 2,
+            "num_warps": 4,
+            "N": 32,
+            "P": 32,
+        },
+    ]:
+        b2b_gemm_template.maybe_append_choice(
+            choices,
+            input_nodes=(mat1, mat2, mat3),
+            layout=layout,
+            **config,
+        )
+    return autotune_select_algorithm("b2b_gemm", choices, [mat1, mat2, mat3], layout)
+
+
+# currently it matches ((A @ B) @ C)
+# TODO: later will change to matching (A @ B) in (epilogue2 ((epilogue1 (A @ B)) @ C)) and inspecting the graph
+# TODO: match more cases such as bmm and addmm
+@register_graph_pattern(
+    CallFunction(aten.mm, CallFunction(aten.mm, Arg(), Arg()), Arg()),
+    pass_dict=B2B_GEMM_PASS,
+)
+def b2b_gemm(
+    match: Match, mat1: torch.fx.Node, mat2: torch.fx.Node, mat3: torch.fx.Node
+) -> None:
+    if can_apply_b2b_gemm(mat1, mat2, mat3):
```

Comment:
I think it's slightly better to do this as an extra_check: https://github.com/pytorch/pytorch/blob/e1499f6342019df7f28784382a5c0920adb1fe73/torch/_inductor/pattern_matcher.py#L1574

Reply:
Changed to extra_check.